File: planner.py
from __future__ import annotations

from datetime import date
from pathlib import Path
from typing import Any

import pandas as pd


MANUAL_SESSION_LABEL = "MANUAL BUNK"


def _normalize_whitespace(value: str) -> str:
    return " ".join(str(value).split()).strip()


def _format_session_time_text(session_time: str) -> str:
    return _normalize_whitespace(str(session_time).upper())


def _extract_session_start(session_time: str) -> pd.Timestamp | None:
    text = str(session_time).split("-", maxsplit=1)[0].strip().upper()
    if not text:
        return None

    parsed = pd.to_datetime(text, format="%I:%M%p", errors="coerce")
    if pd.isna(parsed):
        parsed = pd.to_datetime(text, format="%I%p", errors="coerce")
    if pd.isna(parsed):
        return None
    return parsed


def _match_event(candidate_date: date, event_dates: list[date], lookback_days: int) -> tuple[date, int] | None:
    matches: list[tuple[date, int]] = []

    for event_date in sorted(set(event_dates)):
        days_before = (event_date - candidate_date).days
        if 0 <= days_before <= lookback_days:
            matches.append((event_date, days_before))

    if not matches:
        return None

    return min(matches, key=lambda item: (item[1], item[0]))


def build_course_catalog(attendance_df: pd.DataFrame) -> pd.DataFrame:
    columns = ["course_code", "subject_name", "faculty", "faculty_email", "course"]
    if attendance_df.empty:
        return pd.DataFrame(columns=columns)

    catalog = (
        attendance_df[columns]
        .dropna(subset=["course_code"])
        .drop_duplicates(subset=["course_code"])
        .sort_values(by=["course_code"])
        .reset_index(drop=True)
    )
    return catalog


def build_manual_bunks(manual_entries: pd.DataFrame, course_catalog: pd.DataFrame) -> pd.DataFrame:
    expected_columns = [
        "period_date",
        "session_time",
        "course_code",
        "subject_name",
        "faculty",
        "faculty_email",
        "course",
        "score",
        "source",
    ]
    if manual_entries.empty:
        return pd.DataFrame(columns=expected_columns)

    merged = manual_entries.copy()
    merged = merged.dropna(subset=["date", "course_code"])
    merged["course_code"] = merged["course_code"].map(_normalize_whitespace)
    merged = merged.loc[merged["course_code"] != ""].copy()
    if merged.empty:
        return pd.DataFrame(columns=expected_columns)

    merged["date"] = pd.to_datetime(merged["date"], errors="coerce")
    merged = merged.loc[merged["date"].notna()].copy()
    if merged.empty:
        return pd.DataFrame(columns=expected_columns)

    catalog_lookup = course_catalog[
        ["course_code", "subject_name", "faculty", "faculty_email", "course"]
    ].drop_duplicates(subset=["course_code"])
    merged = merged.merge(catalog_lookup, on="course_code", how="left")
    merged["session_time"] = merged["session_time"].fillna("").map(_normalize_whitespace)
    merged.loc[merged["session_time"] == "", "session_time"] = MANUAL_SESSION_LABEL
    merged["period_date"] = merged["date"]
    merged["score"] = "manual"
    merged["source"] = "manual"

    return merged[expected_columns].copy()


def build_not_marked_candidates(not_marked_entries: pd.DataFrame) -> pd.DataFrame:
    expected_columns = [
        "period_date",
        "session_time",
        "course_code",
        "subject_name",
        "faculty",
        "faculty_email",
        "course",
        "score",
        "source",
    ]
    if not_marked_entries.empty:
        return pd.DataFrame(columns=expected_columns)

    selected = not_marked_entries.copy()
    selected["period_date"] = pd.to_datetime(selected["period_date"], errors="coerce")
    selected = selected.loc[selected["period_date"].notna()].copy()
    if selected.empty:
        return pd.DataFrame(columns=expected_columns)

    selected["score"] = "not_marked"
    selected["source"] = "not_marked"
    return selected[expected_columns].copy()


def generate_duty_leave_plan(
    attendance_df: pd.DataFrame,
    event_dates: list[date],
    course_limits: dict[str, int],
    *,
    cutoff_date: date | None = None,
    lookback_days: int = 4,
    manual_entries: pd.DataFrame | None = None,
    not_marked_entries: pd.DataFrame | None = None,
) -> pd.DataFrame:
    candidate_columns = [
        "period_date",
        "session_time",
        "course_code",
        "subject_name",
        "faculty",
        "faculty_email",
        "course",
        "score",
        "source",
    ]
    if attendance_df.empty:
        return pd.DataFrame(columns=candidate_columns)

    lms_candidates = attendance_df.loc[attendance_df["score"] == "0/1"].copy()
    lms_candidates["source"] = "lms"
    lms_candidates = lms_candidates[candidate_columns]

    catalog = build_course_catalog(attendance_df)
    manual_candidates = build_manual_bunks(
        manual_entries if manual_entries is not None else pd.DataFrame(),
        catalog,
    )
    not_marked_candidates = build_not_marked_candidates(
        not_marked_entries if not_marked_entries is not None else pd.DataFrame()
    )

    combined = pd.concat(
        [lms_candidates, manual_candidates, not_marked_candidates],
        ignore_index=True,
    )
    combined["period_date"] = pd.to_datetime(combined["period_date"], errors="coerce")
    combined = combined.loc[combined["period_date"].notna()].copy()

    if cutoff_date is not None:
        combined = combined.loc[combined["period_date"].dt.date <= cutoff_date].copy()

    combined["session_start"] = combined["session_time"].map(_extract_session_start)
    combined["course_limit"] = combined["course_code"].map(course_limits).fillna(0).astype(int)
    combined = combined.loc[combined["course_limit"] > 0].copy()
    if combined.empty:
        return combined

    if event_dates:
        combined["event_match"] = combined["period_date"].dt.date.map(
            lambda candidate_date: _match_event(candidate_date, event_dates, lookback_days)
        )
        combined = combined.loc[combined["event_match"].notna()].copy()
        if combined.empty:
            return combined

        combined["matched_event_date"] = combined["event_match"].map(lambda item: item[0])
        combined["days_before_event"] = combined["event_match"].map(lambda item: item[1])
        combined = combined.sort_values(
            by=[
                "course_code",
                "days_before_event",
                "matched_event_date",
                "period_date",
                "session_start",
                "source",
                "subject_name",
            ],
            na_position="last",
        )
    else:
        combined["matched_event_date"] = pd.NaT
        combined["days_before_event"] = pd.NA
        combined = combined.sort_values(
            by=["course_code", "period_date", "session_start", "source", "subject_name"],
            na_position="last",
        )

    selected_groups: list[pd.DataFrame] = []
    for course_code, group in combined.groupby("course_code", sort=False):
        limit = int(group["course_limit"].iloc[0])
        selected_groups.append(group.head(limit))

    selected = pd.concat(selected_groups, ignore_index=True) if selected_groups else combined.iloc[0:0].copy()
    selected = selected.sort_values(
        by=["period_date", "session_start", "course_code", "subject_name", "source"],
        na_position="last",
    ).reset_index(drop=True)
    selected["date"] = selected["period_date"].dt.strftime("%d-%m-%Y")

    return selected


def recommendation_preview(selected_df: pd.DataFrame) -> pd.DataFrame:
    if selected_df.empty:
        return pd.DataFrame(
            columns=[
                "date",
                "session_time",
                "course",
                "faculty",
                "faculty_email",
                "source",
                "matched_event_date",
                "days_before_event",
            ]
        )

    preview = selected_df.copy()
    preview["course"] = preview["course_code"] + " " + preview["subject_name"]
    preview["matched_event_date"] = (
        pd.to_datetime(preview["matched_event_date"], errors="coerce")
        .dt.strftime("%d-%m-%Y")
        .fillna("")
    )
    preview["days_before_event"] = preview["days_before_event"].fillna("")
    return preview[
        [
            "date",
            "session_time",
            "course",
            "faculty",
            "faculty_email",
            "source",
            "matched_event_date",
            "days_before_event",
        ]
    ]


def format_daywise_text(selected_df: pd.DataFrame, *, append_counts: bool = True) -> str:
    if selected_df.empty:
        return ""

    lines: list[str] = []
    grouped = selected_df.groupby("date", sort=False)

    for index, (date_text, group) in enumerate(grouped):
        if index > 0:
            lines.append("----")
        lines.append(date_text)

        for row in group.itertuples(index=False):
            lines.append(
                f"{_format_session_time_text(row.session_time)} : "
                f"{row.course_code} : {row.subject_name} : {row.faculty} : {row.faculty_email}"
            )

    if append_counts:
        lines.append("----")
        lines.append("DL Count By Course")
        counts = (
            selected_df.groupby(["course_code", "subject_name"], sort=True)
            .size()
            .reset_index(name="count")
            .sort_values(by=["course_code", "subject_name"])
        )
        for row in counts.itertuples(index=False):
            lines.append(f"{row.course_code} : {row.subject_name} : {row.count}")

    return "\n".join(lines) + "\n"


def write_plan_outputs(selected_df: pd.DataFrame, text_path: Path, csv_path: Path) -> None:
    preview = recommendation_preview(selected_df)
    preview.to_csv(csv_path, index=False)
    text_path.write_text(format_daywise_text(selected_df), encoding="utf-8")

File: frontend/lib/planner.ts
import type {
  AttendanceRow,
  CourseLimitInput,
  ManualBunkInput,
  PlannerCourseCount,
  PlannerGenerateRequest,
  PlannerGenerateResponse,
  PlannerRow,
} from "@/lib/types"

type CourseCatalogEntry = {
  course_code: string
  subject_name: string
  faculty: string
  faculty_email: string
  course: string
}

type PlannerCandidate = {
  period_date: string
  session_time: string
  course_code: string
  subject_name: string
  faculty: string
  faculty_email: string
  course: string
  score: string
  source: "lms" | "manual" | "not_marked"
  session_start: number | null
  course_limit: number
  matched_event_date: string | null
  days_before_event: number | null
}

const MANUAL_SESSION_LABEL = "MANUAL BUNK"

function normalizeWhitespace(value: string | null | undefined) {
  return (value || "").split(/\s+/).filter(Boolean).join(" ").trim()
}

function formatSessionTimeText(value: string | null | undefined) {
  return normalizeWhitespace(value).toUpperCase()
}

function isoToDisplayDate(value: string | null | undefined) {
  if (!value) {
    return ""
  }

  const match = /^(\d{4})-(\d{2})-(\d{2})$/.exec(value)
  if (!match) {
    return value
  }

  return `${match[3]}-${match[2]}-${match[1]}`
}

function parseIsoDate(value: string) {
  const match = /^(\d{4})-(\d{2})-(\d{2})$/.exec(value)
  if (!match) {
    return null
  }

  return Date.UTC(Number(match[1]), Number(match[2]) - 1, Number(match[3]))
}

function daysBetween(eventDate: string, candidateDate: string) {
  const eventTime = parseIsoDate(eventDate)
  const candidateTime = parseIsoDate(candidateDate)
  if (eventTime === null || candidateTime === null) {
    return null
  }

  return Math.round((eventTime - candidateTime) / 86_400_000)
}

function extractSessionStart(sessionTime: string | null | undefined) {
  const text = normalizeWhitespace(sessionTime).split("-", 1)[0]?.toUpperCase() || ""
  const match = /^(\d{1,2})(?::(\d{2}))?\s*(AM|PM)$/.exec(text)
  if (!match) {
    return null
  }

  let hours = Number(match[1]) % 12
  const minutes = Number(match[2] || "0")
  if (match[3] === "PM") {
    hours += 12
  }

  return hours * 60 + minutes
}

function buildCourseCatalog(attendanceRows: AttendanceRow[]) {
  const seen = new Set<string>()
  const catalog: CourseCatalogEntry[] = []

  for (const row of attendanceRows) {
    const courseCode = normalizeWhitespace(row.course_code)
    if (!courseCode || seen.has(courseCode)) {
      continue
    }

    seen.add(courseCode)
    catalog.push({
      course_code: courseCode,
      subject_name: normalizeWhitespace(row.subject_name),
      faculty: normalizeWhitespace(row.faculty),
      faculty_email: normalizeWhitespace(row.faculty_email),
      course: normalizeWhitespace(row.course),
    })
  }

  return catalog.sort((left, right) => left.course_code.localeCompare(right.course_code))
}

function buildManualCandidates(
  manualEntries: ManualBunkInput[],
  catalogByCourse: Map<string, CourseCatalogEntry>
) {
  return manualEntries
    .map((entry) => {
      const courseCode = normalizeWhitespace(entry.course_code)
      const periodDate = normalizeWhitespace(entry.date)
      if (!courseCode || !periodDate) {
        return null
      }

      const course = catalogByCourse.get(courseCode)
      const sessionTime = normalizeWhitespace(entry.session_time) || MANUAL_SESSION_LABEL

      return {
        period_date: periodDate,
        session_time: sessionTime,
        course_code: courseCode,
        subject_name: course?.subject_name || "",
        faculty: course?.faculty || "",
        faculty_email: course?.faculty_email || "",
        course: course?.course || "",
        score: "manual",
        source: "manual" as const,
      }
    })
    .filter((entry): entry is NonNullable<typeof entry> => Boolean(entry))
}

function buildNotMarkedCandidates(
  attendanceRows: AttendanceRow[],
  selectedRecordIds: string[]
) {
  const selected = new Set(selectedRecordIds)

  return attendanceRows
    .filter((row) => selected.has(row.record_id))
    .map((row) => ({
      period_date: normalizeWhitespace(row.period_date),
      session_time: normalizeWhitespace(row.session_time),
      course_code: normalizeWhitespace(row.course_code),
      subject_name: normalizeWhitespace(row.subject_name),
      faculty: normalizeWhitespace(row.faculty),
      faculty_email: normalizeWhitespace(row.faculty_email),
      course: normalizeWhitespace(row.course),
      score: "not_marked",
      source: "not_marked" as const,
    }))
    .filter((row) => row.period_date)
}

function compareNullableNumbers(left: number | null, right: number | null) {
  if (left === right) {
    return 0
  }
  if (left === null) {
    return 1
  }
  if (right === null) {
    return -1
  }
  return left - right
}

function compareNullableStrings(left: string | null, right: string | null) {
  if (left === right) {
    return 0
  }
  if (!left) {
    return 1
  }
  if (!right) {
    return -1
  }
  return left.localeCompare(right)
}

function compareCandidates(left: PlannerCandidate, right: PlannerCandidate) {
  return (
    left.period_date.localeCompare(right.period_date) ||
    compareNullableNumbers(left.session_start, right.session_start) ||
    left.course_code.localeCompare(right.course_code) ||
    left.subject_name.localeCompare(right.subject_name) ||
    left.source.localeCompare(right.source)
  )
}

function buildCourseCounts(selected: PlannerCandidate[]) {
  const counts = new Map<string, PlannerCourseCount>()

  for (const row of selected) {
    const key = `${row.course_code}:::${row.subject_name}`
    const existing = counts.get(key)
    if (existing) {
      existing.count += 1
      continue
    }

    counts.set(key, {
      course_code: row.course_code || null,
      subject_name: row.subject_name || null,
      count: 1,
    })
  }

  return [...counts.values()].sort((left, right) => {
    return (
      compareNullableStrings(left.course_code, right.course_code) ||
      compareNullableStrings(left.subject_name, right.subject_name)
    )
  })
}

function buildPlannerRows(selected: PlannerCandidate[]): PlannerRow[] {
  return selected.map((row) => ({
    date: isoToDisplayDate(row.period_date),
    session_time: row.session_time || null,
    course: normalizeWhitespace(`${row.course_code} ${row.subject_name}`) || null,
    faculty: row.faculty || null,
    faculty_email: row.faculty_email || null,
    source: row.source,
    matched_event_date: row.matched_event_date
      ? isoToDisplayDate(row.matched_event_date)
      : "",
    days_before_event: row.days_before_event ?? "",
  }))
}

function formatDaywiseText(selected: PlannerCandidate[], courseCounts: PlannerCourseCount[]) {
  if (!selected.length) {
    return ""
  }

  const lines: string[] = []
  let currentDate = ""

  for (const row of selected) {
    const dateText = isoToDisplayDate(row.period_date)
    if (dateText !== currentDate) {
      if (currentDate) {
        lines.push("----")
      }
      currentDate = dateText
      lines.push(currentDate)
    }

    lines.push(
      `${formatSessionTimeText(row.session_time)} : ${row.course_code} : ${row.subject_name} : ${row.faculty} : ${row.faculty_email}`
    )
  }

  lines.push("----")
  lines.push("DL Count By Course")
  for (const count of courseCounts) {
    lines.push(`${count.course_code} : ${count.subject_name} : ${count.count}`)
  }

  return `${lines.join("\n")}\n`
}

export function generatePlannerResponse(
  request: PlannerGenerateRequest
): PlannerGenerateResponse {
  const attendanceRows = request.attendance_rows || []
  if (!attendanceRows.length) {
    throw new Error("attendance_rows are required for the standalone planner route.")
  }

  const catalog = buildCourseCatalog(attendanceRows)
  const catalogByCourse = new Map(
    catalog.map((entry) => [entry.course_code, entry] as const)
  )

  const lmsCandidates = attendanceRows
    .filter((row) => row.score === "0/1")
    .map((row) => ({
      period_date: normalizeWhitespace(row.period_date),
      session_time: normalizeWhitespace(row.session_time),
      course_code: normalizeWhitespace(row.course_code),
      subject_name: normalizeWhitespace(row.subject_name),
      faculty: normalizeWhitespace(row.faculty),
      faculty_email: normalizeWhitespace(row.faculty_email),
      course: normalizeWhitespace(row.course),
      score: normalizeWhitespace(row.score),
      source: "lms" as const,
    }))
    .filter((row) => row.period_date)

  const manualCandidates = buildManualCandidates(
    request.manual_entries || [],
    catalogByCourse
  )
  const notMarkedCandidates = buildNotMarkedCandidates(
    attendanceRows,
    request.not_marked_record_ids || []
  )

  const courseLimits = new Map(
    (request.course_limits || []).map((item: CourseLimitInput) => [
      item.course_code,
      Number(item.max_dl || 0),
    ])
  )

  let combined: PlannerCandidate[] = [
    ...lmsCandidates,
    ...manualCandidates,
    ...notMarkedCandidates,
  ]
    .filter((row) => row.period_date)
    .map((row) => ({
      ...row,
      session_start: extractSessionStart(row.session_time),
      course_limit: courseLimits.get(row.course_code) ?? 0,
      matched_event_date: null,
      days_before_event: null,
    }))
    .filter((row) => row.course_limit > 0)

  if (request.cutoff_date) {
    combined = combined.filter((row) => row.period_date <= request.cutoff_date!)
  }

  if (request.event_dates.length) {
    const matchedCandidates = combined
      .map((row): (PlannerCandidate & { matched_event_date: string; days_before_event: number }) | null => {
        const matches = request.event_dates
          .map((eventDate) => {
            const daysBefore = daysBetween(eventDate, row.period_date)
            if (daysBefore === null || daysBefore < 0 || daysBefore > request.lookback_days) {
              return null
            }

            return {
              eventDate,
              daysBefore,
            }
          })
          .filter(
            (match): match is { eventDate: string; daysBefore: number } => Boolean(match)
          )
          .sort((left, right) => {
            return left.daysBefore - right.daysBefore || left.eventDate.localeCompare(right.eventDate)
          })

        if (!matches.length) {
          return null
        }

        return {
          ...row,
          matched_event_date: matches[0].eventDate,
          days_before_event: matches[0].daysBefore,
        }
      })
      .filter(
        (
          row
        ): row is PlannerCandidate & {
          matched_event_date: string
          days_before_event: number
        } => row !== null
      )
      .sort((left, right) => {
        return (
          left.course_code.localeCompare(right.course_code) ||
          (left.days_before_event ?? 99) - (right.days_before_event ?? 99) ||
          compareNullableStrings(left.matched_event_date, right.matched_event_date) ||
          left.period_date.localeCompare(right.period_date) ||
          compareNullableNumbers(left.session_start, right.session_start) ||
          left.source.localeCompare(right.source) ||
          left.subject_name.localeCompare(right.subject_name)
        )
      })

    combined = matchedCandidates
  } else {
    combined = combined.sort((left, right) => {
      return (
        left.course_code.localeCompare(right.course_code) ||
        left.period_date.localeCompare(right.period_date) ||
        compareNullableNumbers(left.session_start, right.session_start) ||
        left.source.localeCompare(right.source) ||
        left.subject_name.localeCompare(right.subject_name)
      )
    })
  }

  const selected: PlannerCandidate[] = []
  const takenByCourse = new Map<string, number>()
  for (const row of combined) {
    const taken = takenByCourse.get(row.course_code) ?? 0
    if (taken >= row.course_limit) {
      continue
    }
    takenByCourse.set(row.course_code, taken + 1)
    selected.push(row)
  }

  selected.sort(compareCandidates)

  const plannerRows = buildPlannerRows(selected)
  const courseCounts = buildCourseCounts(selected)
  const plannerText = formatDaywiseText(selected, courseCounts)
  const plannerCsv = plannerRows.length
    ? [
        [
          "date",
          "session_time",
          "course",
          "faculty",
          "faculty_email",
          "source",
          "matched_event_date",
          "days_before_event",
        ].join(","),
        ...plannerRows.map((row) =>
          [
            row.date ?? "",
            row.session_time ?? "",
            row.course ?? "",
            row.faculty ?? "",
            row.faculty_email ?? "",
            row.source ?? "",
            row.matched_event_date ?? "",
            row.days_before_event ?? "",
          ]
            .map((value) => `"${String(value).replaceAll('"', '""')}"`)
            .join(",")
        ),
      ].join("\n")
    : ""

  return {
    summary: {
      recommended_rows: selected.length,
      courses_covered: new Set(selected.map((row) => row.course_code)).size,
      manual_or_not_marked_used: selected.filter((row) =>
        ["manual", "not_marked"].includes(row.source)
      ).length,
    },
    planner_rows: plannerRows,
    course_counts: courseCounts,
    planner_text: plannerText,
    planner_csv: plannerCsv,
  }
}
